### PR TITLE
Deepspeed integration for 7B models

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR attempts to bring deepspeed integration to empower tuning with 7B models. In the [summarize from human feedback paper](https://arxiv.org/pdf/2009.01325.pdf), the experimented with 1.3B, 2.7B, and 6.7B models, so this PR would in principle allow us to replicate that work. 

Some of the notable changes needed to make things work:
* `mixed_precision: 'bf16'` turns out to be important, otherwise OOM.
* Initialize all models on CPU first and use ` accelerator.prepare` and `deepspeed.initialize`, otherwise OOM.
* Enable `bf16` for `reward_model` and `ref_policy`, otherwise OOM.
* Do not log the histogram of ratio, otherwise OOM.
* Enable gradient checkpointing, otherwise OOM.
* In https://github.com/OpenLMLab/MOSS-RLHF/blob/40b91eb2f2b71b16919addede0341d2bef70825d/utils.py#L41-L43, they have an additional `critic_model`, which they finally offload `reward_model`, `critic_model`, and `ref_policy` to CPU, but it is not necessary in our case.

Here is a training run  https://wandb.ai/costa-huang/cleanRL/runs/kve7tu43/overview with 

```
accelerate launch --config_file deepspeed.yaml lm_human_preference_details/train_policy_accelerate.py \
    --rewards.trained_model ''  \
    --base_model tiiuae/falcon-7b  \
    --no_use_tensorflow_adam \
    --ppo.gradient_accumulation_steps 64 \
    --track
```

Training results was pretty bad, but I think this is probably some issue related to model compatibility. To replicate [summarize from human feedback paper](https://arxiv.org/pdf/2009.01325.pdf), we should probably use the OPT models which have 1.3B, 2.7B, and 6.7B models.

CC @lewtun

### CODE REVIEW

File: lm_human_preference_details/train_policy_accelerate.py [hunk 1]
```diff
@@ -510,6 +514,37 @@ def train(args: Args):
     )
     dataloader = DataLoader(dataset, batch_size=args.ppo.local_batch_size)
     policy, optimizer, dataloader = accelerator.prepare(policy, optimizer, dataloader)
+    if args.deepspeed:
+        import deepspeed
+
+        deepspeed_states = AcceleratorState().deepspeed_plugin
+        deepspeed_states.deepspeed_config['train_micro_batch_size_per_gpu'] = args.ppo.local_micro_batch_size
+        deepspeed_states.deepspeed_config['checkpoint'] = {'use_node_local_storage': True}
+        off_load_device = "cpu"
```
Comment (lewtun) [inline lm_human_preference_details/train_policy_accelerate.py]: Note that this will slow down your code significantly. I would allow the option to be set as an option that's inferred from the `accelerate` config as I did here: https://github.com/huggingface/trl/pull/758

File: lm_human_preference_details/train_policy_accelerate.py [hunk 2]
```diff
@@ -510,6 +514,37 @@ def train(args: Args):
     )
     dataloader = DataLoader(dataset, batch_size=args.ppo.local_batch_size)
     policy, optimizer, dataloader = accelerator.prepare(policy, optimizer, dataloader)
+    if args.deepspeed:
+        import deepspeed
+
+        deepspeed_states = AcceleratorState().deepspeed_plugin
+        deepspeed_states.deepspeed_config['train_micro_batch_size_per_gpu'] = args.ppo.local_micro_batch_size
+        deepspeed_states.deepspeed_config['checkpoint'] = {'use_node_local_storage': True}
```
Comment (lewtun) [inline lm_human_preference_details/train_policy_accelerate.py]: I think this flag is only needed if each node has a separate local filesystem. For the HFC case, you probably don't need it

File: lm_human_preference_details/train_policy_accelerate.py [hunk 3]
```diff
@@ -510,6 +514,37 @@ def train(args: Args):
     )
     dataloader = DataLoader(dataset, batch_size=args.ppo.local_batch_size)
     policy, optimizer, dataloader = accelerator.prepare(policy, optimizer, dataloader)
+    if args.deepspeed:
+        import deepspeed
+
+        deepspeed_states = AcceleratorState().deepspeed_plugin
+        deepspeed_states.deepspeed_config['train_micro_batch_size_per_gpu'] = args.ppo.local_micro_batch_size
```
Comment (lewtun) [inline lm_human_preference_details/train_policy_accelerate.py]: If I'm not mistaken, these config values are set automatically by the accelerator and don't need to be overridden

File: lm_human_preference_details/train_policy_accelerate.py [hunk 4]
```diff
@@ -510,6 +514,37 @@ def train(args: Args):
...
+            #     "stage": stage,
+            #     "stage3_param_persistence_threshold": 1e4,
+            #     "offload_param": {
+            #         "device": off_load_device
+            #     }
+            # },
+            "bf16": {
+                "enabled": True
+            },
+            "prescale_gradients": False,
```
Comment (lewtun) [inline lm_human_preference_details/train_policy_accelerate.py:546]: I think this flag and the one below are false by default, so probably don't need to be set either

File: lm_human_preference_details/train_policy_accelerate.py [hunk 5]
```diff
@@ -755,7 +790,8 @@ def train(args: Args):
                     )
 
         with torch.no_grad():
-            writer.add_histogram("ppo/val/ratio_hist", ratio, update)
+            if not args.deepspeed: # for some reason there is a OOM with the `writer.add_histogram`
+                writer.add_histogram("ppo/val/ratio_hist", ratio, update)
```
Comment (lewtun) [inline lm_human_preference_details/train_policy_accelerate.py:786]: FYI I was able to train 7B models in TRL with ZeRO-2 and didn't need to remove the histogram. On the other hand that was for sentiment tuning, which is less memory intensive than your application here

**Comment by vwxyzjn** [comment]
Confirmed that it can reasonably run 7b models (no benchmark results yet)

```

SAVE_PATH_REWARD="models/train_7b_$(date +%s)/reward.pt"
SAVE_PATH_POLICY="models/train_7b_$(date +%s)/policy.pt"
poetry run accelerate launch --config_file deepspeed.yaml  lm_human_preference_details/train_reward_accelerate.py \
    --base_model cerebras/Cerebras-GPT-6.7B \
    --no_use_tensorflow_adam \
    --gradient_accumulation_steps=4 \
    --local_rollout_batch_size=4 \
    --save_path=$SAVE_PATH_REWARD \
    --track && \
    poetry run accelerate launch --config_file deepspeed.yaml  lm_human_preference_details/train_policy_accelerate.py \
    --rewards.trained_model=$SAVE_PATH_REWARD \
    --base_model=cerebras/Cerebras-GPT-6.7B \
    --deepspeed \
    --no_use_tensorflow_adam \
    --ppo.gradient_accumulation_steps 64 \
    --track
```

https://wandb.ai/costa-huang/cleanRL/runs/hn9wtka9?workspace=user-costa-huang

image
